### PR TITLE
[#21515][Go SDK] Update go protobuf package to new version

### LINKED ISSUE #21515: Update Go protobuf package to new version
The original Go protobuf package (github.com/golang/protobuf/proto) is deprecated and should be replaced with the new version (google.golang.org/protobuf/proto)

Imported from Jira [BEAM-14242](https://issues.apache.org/jira/browse/BEAM-14242). Original Jira may contain additional context.
Reported by: jrmccluskey.
Subtask of issue #21513

### PR DESCRIPTION
**Pull request by imvtsl**
Update go protobuf package to new version.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] fixes #21515.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Review by lostluck — CHANGES_REQUESTED** [review]
Similarly, for `create_test.go` the code there is validating that a ProtoV1 message is supported and can be encoded.

You can remove the marshal interface lines.

	_ proto.Marshaler   = &testProto{}
	_ proto.Unmarshaler = &testProto{}
	
Since we aren't relying on the protoV1 unmarshalling handlers after this PR.

**Comment by imvtsl** [comment]
retest this please

File: sdks/go/pkg/beam/coder.go [hunk 1]
```diff
@@ -51,7 +50,7 @@ type jsonCoder interface {
 	json.Unmarshaler
 }
 
-var protoMessageType = reflect.TypeOf((*protov1.Message)(nil)).Elem()
```
[the remark this answers is not in the document]
Comment (imvtsl) [reply sdks/go/pkg/beam/coder.go:54]: Thank you for informing about protoadapt. I had missed it.
I implemented this change.

File: sdks/go/pkg/beam/coder.go [hunk 2]
```diff
@@ -51,7 +50,7 @@ type jsonCoder interface {
...
+var protoMessageType = reflect.TypeOf((*protov2.Message)(nil)).Elem()
```
[the remark this answers is not in the document]
Comment (imvtsl) [reply sdks/go/pkg/beam/coder.go]: Thank you for informing about protoadapt and that we want to keep it compatible with clients using v1. This helped a lot.
I implemented this change.

File: sdks/go/pkg/beam/coder.go [hunk 3]
```diff
@@ -293,8 +290,6 @@ func protoDec(t reflect.Type, in []byte) (T, error) {
 	switch it := reflect.New(t.Elem()).Interface().(type) {
 	case protoreflect.ProtoMessage:
 		p = it
-	case protov1.Message:
-		p = protov1.MessageV2(it)
```
[the remark this answers is not in the document]
Comment (imvtsl) [reply sdks/go/pkg/beam/coder.go:297]: done. Thanks.

File: sdks/go/pkg/beam/core/runtime/harness/statemgr.go [hunk 4]
```diff
@@ -633,15 +633,16 @@ func (c *StateChannel) read(ctx context.Context) {
 		if !ok {
 			// This can happen if Send returns an error that write handles, but
 			// the message was actually sent.
-			log.Errorf(ctx, "StateChannel[%v].read: no consumer for state response: %v", c.id, proto.MarshalTextString(msg))
+			bytes, _ := prototext.Marshal(msg)
```
[the remark this answers is not in the document]
Comment (imvtsl) [reply sdks/go/pkg/beam/core/runtime/harness/statemgr.go]: Done. I went with msg.String() approach, as we have used this method in other places to keep uniformity in the codebase.

File: sdks/go/pkg/beam/core/runtime/graphx/translate.go [hunk 5]
```diff
@@ -1209,13 +1209,13 @@ func (m *marshaller) addWindowingStrategy(w *window.WindowingStrategy) (string,
 }
 
 func (m *marshaller) internWindowingStrategy(w *pipepb.WindowingStrategy) string {
-	key := proto.MarshalTextString(w)
-	if id, exists := m.windowing2id[key]; exists {
+	key := w.String()
+	if id, exists := m.windowing2id[string(key)]; exists {
```
[the remark this answers is not in the document]
Comment (imvtsl) [reply sdks/go/pkg/beam/core/runtime/graphx/translate.go]: Yeah, I missed it. Thanks for pointing it out.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go [hunk 6]
```diff
@@ -615,8 +615,8 @@ func (b *CoderMarshaller) internRowCoder(schema *pipepb.Schema) string {
 }
 
 func (b *CoderMarshaller) internCoder(coder *pipepb.Coder) string {
-	key := proto.MarshalTextString(coder)
-	if id, exists := b.coder2id[key]; exists {
+	key := coder.String()
+	if id, exists := b.coder2id[string(key)]; exists {
```
[the remark this answers is not in the document]
Comment (imvtsl) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Yeah, I missed it. Thanks for pointing it out.

File: sdks/go/pkg/beam/coder.go [hunk 7]
```diff
@@ -276,8 +275,6 @@ func protoEnc(in T) ([]byte, error) {
 	switch it := in.(type) {
 	case protoreflect.ProtoMessage:
 		p = it
-	case protov1.Message:
-		p = protov1.MessageV2(it)
```
[the remark this answers is not in the document]
Comment (imvtsl) [reply sdks/go/pkg/beam/coder.go:280]: Done.

**Comment by imvtsl** [comment]
> Similarly, for `create_test.go` the code there is validating that a ProtoV1 message is supported and can be encoded.
> 

Thank you!! I had no idea about protoadapt package. I was trying to modify tests to somehow work with proto v2 message. This helped clarified the goal of the test case.

> You can remove the marshal interface lines.
> 
> ```
> _ proto.Marshaler   = &testProto{}
> _ proto.Unmarshaler = &testProto{}
> ```
> 
> Since we aren't relying on the protoV1 unmarshalling handlers after this PR.

Done.

**Comment by imvtsl** [comment]
@lostluck 
Although, this change doesn't impact any users, I believe I still have to add an entry in [Changes.md ](https://github.com/apache/beam/blob/master/CHANGES.md). Do I add the entry under section 2.59.0 or 2.58.0?

**Comment by lostluck** [comment]
It would go under 2.59.0 as 2.58 has already been cut and is going to be released next week.

I agree that the change is worth a small note, in the unlikely event that it breaks a protov1 user.

**Comment by imvtsl** [comment]
> It would go under 2.59.0 as 2.58 has already been cut and is going to be released next week.
> 
> I agree that the change is worth a small note, in the unlikely event that it breaks a protov1 user.

Thanks. I pushed the change.

File: CHANGES.md [hunk 8]
```diff
@@ -80,6 +80,7 @@
 ## Bugfixes
 
 * Fixed incorrect service account impersonation flow for Python pipelines using BigQuery IOs ([#32030](https://github.com/apache/beam/issues/32030)).
+* Updated Go protobuf package to new version (Go) ([#21515](https://github.com/apache/beam/issues/21515)).
```
Comment (lostluck) [inline CHANGES.md:83]: I'd put this under "Improvements" it's not fixing any bugs. It's removing use of deprecated code (that under the hood uses the v2 packages anyway).

Comment (imvtsl) [reply CHANGES.md]: Done.

**Review by lostluck — APPROVED** [review]
LGTM
Thank you very much! This is very helpful.

**Comment by imvtsl** [comment]
Thank you!!

**Pull request merged**